internal/backend/remote-state/http/client_test.go

```diff
@@ -251,3 +253,93 @@ func TestHttpClient_IsLockingEnabled(t *testing.T) {
 		})
 	}
 }
+
+// Tests the Lock method for the HTTP client.
+// Test to see correct lock info is returned
+func TestHttpClient_lock(t *testing.T) {
+	stateLockInfoA := statemgr.LockInfo{
+		ID:        "ada-lovelace-state-lock-id",
+		Who:       "AdaLovelace",
+		Operation: "TestTypePlan",
+		Created:   time.Date(2023, time.August, 16, 15, 9, 26, 0, time.UTC),
+	}
+
+	testCases := []struct {
+		name                string
+		lockInfo            *statemgr.LockInfo
+		lockResponseStatus  int
+		lockResponseBody    []byte
+		expectedStateLockID string
+		expectedErrorMsg    error
+	}{
+		{
+			// Successful locking HTTP remote state
+			name:                "Successfully locked",
+			lockInfo:            &stateLockInfoA,
+			lockResponseStatus:  http.StatusOK,
+			lockResponseBody:    nil,
+			expectedStateLockID: stateLockInfoA.ID,
+			expectedErrorMsg:    nil,
+		},
+		{
+			// Failed to lock state, HTTP remote state already locked
+			name:                "Locked remote state",
+			lockInfo:            &stateLockInfoA,
+			lockResponseStatus:  http.StatusLocked,
+			lockResponseBody:    []byte(`{"ID":"linus-torvalds-http-remote-state-lock-id", "Path": "", "Operation": "TestTypePlan", "Who": "LinusTorvalds", "Created": "2024-08-15T09:00:26.00Z" }`),
+			expectedStateLockID: "",
+			expectedErrorMsg: &statemgr.LockError{
+				Info: &statemgr.LockInfo{
+					ID:        "linus-torvalds-http-remote-state-lock-id",
+					Who:       "LinusTorvalds",
+					Operation: "TestTypePlan",
+					Created:   time.Date(2024, time.August, 15, 9, 0, 26, 0, time.UTC),
+				},
+				Err: fmt.Errorf("HTTP remote state already locked: ID=%s", "linus-torvalds-http-remote-state-lock-id"),
+			},
+		},
+		{
+			// Failed to lock state HTTP remote state already locked. No remote lock details returned
+			name:                "Locked remote state failed to unmarshal body",
+			lockInfo:            &stateLockInfoA,
+			lockResponseStatus:  http.StatusLocked,
+			lockResponseBody:    nil,
+			expectedStateLockID: "",
+			expectedErrorMsg: &statemgr.LockError{
+				Info: &stateLockInfoA,
+				Err:  fmt.Errorf("HTTP remote state already locked, failed to unmarshal body"),
+			},
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			handler := func(w http.ResponseWriter, _ *http.Request) {
+				w.WriteHeader(tt.lockResponseStatus)
+				w.Write(tt.lockResponseBody)
+			}
+
+			ts := httptest.NewServer(http.HandlerFunc(handler))
+			defer ts.Close()
+
+			lockURL, err := url.Parse(ts.URL)
+			if err != nil {
+				t.Fatalf("Failed to parse lockURL: %v", err)
+			}
+
+			client := &httpClient{
+				LockURL:    lockURL,
+				LockMethod: "LOCK",
+				Client:     retryablehttp.NewClient(),
+			}
+
+			lockID, err := client.Lock(tt.lockInfo)
+			if tt.expectedErrorMsg != nil && err.Error() != tt.expectedErrorMsg.Error() {
+				t.Errorf("Lock() error = %v, want %v", err, tt.expectedErrorMsg)
+			}
```

Comment:
We must check for (1) unexpected error, (2) absence of expected error and (3) mismatched errors. Currently, we only check for (3) and panic on (2).

Reply:
I think we can do it without a separate variable:
```go
			lockID, err := client.Lock(tt.lockInfo)
			if tt.expectedErrorMsg != nil && err == nil {
				// no expected error
			}
			if tt.expectedErrorMsg == nil && err != nil {
				// unexpected error
			}
			if tt.expectedErrorMsg != nil && err != nil && err.Error() != tt.expectedErrorMsg.Error() {
				// mismatched errors
			}			
```

Reply:
I think this capture all use cases 

`if (err != nil) != (tt.expectedErrorMsg != nil) || (err != nil && err.Error() != tt.expectedErrorMsg.Error()) {
	// One of errors is nil and the other is not OR if both are not nil but their messages differ
}
`

Reply:
I would prefer a simpler-to-read option with multiple ifs, it also allows us to create different error messages based on different cases.

Reply:
Hi @ollevche I have updated the tests base on your suggestions, thanks